examples/js/loaders/GLTFLoader.js

```diff
@@ -538,6 +538,146 @@ THREE.GLTFLoader = ( function () {
 	 *
 	 * Specification: https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_pbrSpecularGlossiness
 	 */
+
+	/**
+	 * A sub class of THREE.StandardMaterial with some of the functionality
+	 * changed via the `onBeforeCompile` callback
+	 * @pailhead
+	 */
+
+	function MeshStandardSGMaterial( params ) {
+
+		THREE.MeshStandardMaterial.call( this );
```

Comment:
No problems with `toJSON()` and `THREE.ObjectLoader` for de/serialization on inheriting `MeshStandardMaterial`?

Reply:
Can you make a snippet and describe what you’re asking?

Reply:
Three.js core doesn't know `MeshStandardSGMaterial` so wondering if `THREE.ObjectLoader` deserialize `MeshStandardSGMaterial` correctly. This is one of the reasons why I chose `ShaderMaterial`.

Reply:
I don't think ObjectLoader can handle this correctly either way, since we need the `onBeforeRender` override now. :/

Reply:
Yeah, that's the downside of `onBeforeRender`.

Reply:
I think i would tend to disagree but, im not sure what the problem is. Currently `Material` **has to know about all sub classed materials**, and it obviously doesn't know about this one when just modifying the `GLTFLoader`.

But then again, how did it work with a `ShaderMaterial`?

Reply:
To reiterate:

This being the `GLTF.Loader` (**loader**), it's job should be to consume a `gltf` file and parse it correctly. This i think means making the three scene graph with three materials. 

There are problems with how to accommodate for that extension and there are multiple solutions to this problem. They all involve creating `a material` that would render the correct `effect'?

Either way, because of three's design, a class such as this ends up having too much responsibility. I don't know the specifics of the problem that you're outlining here, but I'm aware of the problems that `Material` system has in general.

With the current solution, i think a lot of the cruft is the code to refresh uniforms. `onBeforeCompile` proves that there is code already that can do this.

Alternatively, one could override the entire `toJSON` function on this material, and probably solve the outlined problem. Again, redundant cruft would be introduced.

I guess this doesn't have to land (now). My main goal of doing this PR is to get stakeholders such as yourselves to be interested in other areas of three's code base. 

One example is https://github.com/mrdoob/three.js/pull/13198, but after having this discussion i'd mention this too https://github.com/mrdoob/three.js/pull/14259. 

Both point to issues with three's core, and both can solve problems of this class. In addition to this they can solve other problems. You are already championing `NodeMaterial` because of `GLTF.Loader`, you could also champion something else.

So:

1. This being a "loader" how is it related to **de**serialization? (in general how are /example loaders related to three.js, what is the paradigm here?)
2. Should `GLTF.Loader` aim to solve everything or just a subset of everything? (like deserializing a generic material?)
3. Can `GLTF.Loader` be used as one of many things to indicate problems in three's core that can be refactored?

Reply:
Yeah, i think you (probably unintentionally) ended blocking up these issues. https://github.com/mrdoob/three.js/issues/14245#issuecomment-397957636

It's interesting to notice how much scrutiny a small PR (this one) goes through, where every line undergoes scrutiny. I think this is how it should be and that code should be peer reviewed. 

This PR, along with others (generic material serialization, declarative material extensions etc) is blocked by a PR that does not undergo the same scrutiny, and that touches 100+ files and 10 **THOUSANDS** lines of code.

https://github.com/mrdoob/three.js/pull/14333/files

Yeah i'm pretty disappointed. I feel that progress is hampered just because of some idea from 2015 that we're still waiting to be executed.

Reply:
I'm in favor of scrutiny. I'm not in favor of blockers. This is like an infinite loop. There are some areas of three's core that **may have issues** and they seem to be set in stone. An experimental evolving API is further fortifying this stone. 

So to reiterate, the (de)serialization of whatever `spec/gloss material` refers to does not work today? Ie. `THREE.ShaderMaterial` can't be (de)serialized? 

If this is so, i would advise to look into other areas of code for these issues.

For example the material (de)serialization issues all stem from breaking some SOLID principles:

https://github.com/mrdoob/three.js/issues/14290

But it's just another issue blocked from being solved by the experimental evolving API known as `NodeMaterial`.

Reply:
This issue is not blocked. @takahirox asked whether serialization works, and mentioned his reasoning for the original implementation, but (so far) there is no suggestion that we reject this PR. I don't see ShaderMaterial mentioned anywhere in `ObjectLoader.js`, so I assume it also cannot be deserialized. Even if it could, GLTFLoader's use case couldn't because we'd need `onBeforeRender` to be preserved.

Reply:
Ok my bad here for flipping out. But also, this was never a place to address the serialization issue.

Reply:
I looked at both `ObjectLoader` and `MaterialLoader`. Both seem to handle `.userData` field:

https://github.com/mrdoob/three.js/blob/master/src/loaders/MaterialLoader.js#L101

Am i doing something wrong if i call this `json within a json`? What problems are you referring to when talking about serialization? If you have some issues to point me to i would very much appreciate it. In the meanwhile i'll try to set up a snippet because i think that this actually does work with `ObjectLoader` just fine :)

Reply:
2 in this case doesn't work. But you can still hack around it 

```
const fakeSlot = look_for_empty_seriazable_map_slot(material)

material.userData.extended.SomeExtension.fakeSpecularSlotName = fakeSlot

const json = material.toJSON()

const obj = parse(json)

extend(material)

obj.material[realSlot] = obj.material[fakeSlot] 
obj.material[fakeSlot] = null
```

You get the idea.

Reply:
```

const material_with_most_texture_slots = THREE.SomeMaterial
const slot_num = 7

serializeArbitraryTextures( texArray ){
 const textureHolder = new THREE.Group()

for ( let i = 0; i < Math.ceil(texArray.length / slot_num) ; i++){
  textureHolder.push( new THREE.Mesh( dummyGeometry, new material_with_most_texture_slots() )
  
  slotList.forEach((slot,j)=>{
    textureHolder[i][slot] = texArray(i * slot_num + j )
  }
}
textureHolder.toJSON() <-- win
```

The point i'm trying to make is that this is not a hack, but a legitimate usage of three.js API. The API is up for debate.